# realtion between accidents and types


In [2]:
import pandas as pd

# Load the data
file_path = 'peshawar.xlsx'
sheet_name = 'accident types'
df = pd.read_excel(file_path, sheet_name=sheet_name)


In [3]:
import statsmodels.api as sm

# Define the dependent and independent variables
X = df[['Derailments', 'Collisions', 'Collisions at LC']]
y = df['No of Accidents']

# Add a constant to the model (intercept)
X = sm.add_constant(X)

# Fit the GLM model
glm_poisson = sm.GLM(y, X, family=sm.families.Poisson()).fit()

# Get the summary of the model
summary = glm_poisson.summary()
print(summary)


                 Generalized Linear Model Regression Results                  
Dep. Variable:        No of Accidents   No. Observations:                    6
Model:                            GLM   Df Residuals:                        2
Model Family:                 Poisson   Df Model:                            3
Link Function:                    Log   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -9.9200
Date:                Wed, 31 Jul 2024   Deviance:                     0.087746
Time:                        16:09:43   Pearson chi2:                   0.0879
No. Iterations:                     4   Pseudo R-squ. (CS):             0.5448
Covariance Type:            nonrobust                                         
                       coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------------
const                0.2693      1.087  

## interpretation

The GLM results reveal the following about how different types of incidents influence the frequency of railway accidents:The GLM results reveal the following about how different types of incidents influence the frequency of railway accidents:

Derailments:
As one can see, an increase in the number of derailing adds approximately 32 of accidents to the expected number of accidents. The second element increased the price by 1% and is equivalent to 1. 321 as a factor. This implies that derailments are catalysts to the number of accidents, coming out as the leading factor of the list. This means that for every level of increase of the derailment incidents, the overall accident frequency will rise dramatically.

Collisions: 
Every new collision increases the number of the shocking accidents by thirty percent. 5% of an amount of a thousand is 1. 295 (1. 295 x Thousand = 1. 295 Thousand). It also presents a fairly high impact in raising the numbers of accidents and is only lower than derailments.

Collisions at Level Crossings:
With each extra collision at a level crossing the number of accidents rises by approximately twenty.  Conclusively, with regards to the accident frequency, collisions that occur at level crossings positively contribute to it but have the smallest score of all the analyzed factors.

In this regard, it can be concluded that the total derailments have the greatest effect on the frequency of accidents being followed by general collision and then collision at level crossing. This implies that safety measures to avoid derailment can be acknowledged to have the greatest potential in lowering the accident rates and that by addressing the collision and especially those at level crossings can also increase safety but less significantly.

# Location

In [8]:
import pandas as pd
import statsmodels.api as sm
from statsmodels.formula.api import glm

# Load the data
file_path = 'peshawar.xlsx'
sheet_name = 'Location of Accidents'
df = pd.read_excel(file_path, sheet_name=sheet_name)

# Replace spaces in column names with underscores
df.columns = df.columns.str.replace(' ', '_')


# Fit the GLM excluding the 'Year' variable
model = glm('No_of_Accidents ~ Accidents_at_Track + Accidents_in_Station_Limits',
            data=df,
            family=sm.families.Poisson()).fit()

# Print the summary of the model
print(model.summary())

                 Generalized Linear Model Regression Results                  
Dep. Variable:        No_of_Accidents   No. Observations:                    6
Model:                            GLM   Df Residuals:                        3
Model Family:                 Poisson   Df Model:                            2
Link Function:                    Log   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -9.9302
Date:                Wed, 31 Jul 2024   Deviance:                      0.10822
Time:                        16:23:19   Pearson chi2:                    0.106
No. Iterations:                     4   Pseudo R-squ. (CS):             0.5433
Covariance Type:            nonrobust                                         
                                  coef    std err          z      P>|z|      [0.025      0.975]
-----------------------------------------------------------------------------------------------
Intercept         

## interpretation
Slightly different from the previous analysis in the GLM tests, the primary concern involves calculating the number of accidents due to the location. The model is fitted using Poisson distribution with a log link function.

Baseline Number of Accidents
If there has been no accident at either site the baseline can be said to be arounn . 572.

The Consequences of Accidents at the Track
The coefficient for Accidents_at_Track variable is 0. 2212. They state that this means that every additional accident at the track results to about a 24 percent increase. Decreased by 8% in the number of total accidents. This result is nevertheless statistically significant which means that accidents at the track indeed are equally important when it comes to the total number of accidents.

The influence of accidents in Station limits
The coefficient of accidental frequency in station limits is 0. 2050. From this it can be inferred that a unit rise of the number of accidents within station limits leads to a rise of about 22. of them 8% in the total of the accidents. Yet this difference might not be statistically considerable, which means that accidents within station boundaries could be less essential than those at the track in defining the overall accidents.

Summary
The findings of this model point to the fact that accidents that occur at the track greatly influence the number of accidents as compared to the influence of accidents that occur within the station limits.

# Type of Line

In [9]:
import pandas as pd
import statsmodels.api as sm
from statsmodels.formula.api import glm

# Load the data
file_path = 'peshawar.xlsx'
sheet_name = 'type of line'
df = pd.read_excel(file_path, sheet_name=sheet_name)

# Replace spaces in column names with underscores
df.columns = df.columns.str.replace(' ', '_')

# Reshape the data to a long format
df_long = pd.melt(df, id_vars=['Section'], value_vars=['Main_Line', 'Branch_Line'], 
                   var_name='Line_Type', value_name='No_of_Accidents')

# Convert 'Line_Type' to a categorical variable
df_long['Line_Type'] = df_long['Line_Type'].astype('category')

# Fit the GLM
model = glm('No_of_Accidents ~ Line_Type', 
            data=df_long,
            family=sm.families.Poisson()).fit()

# Print the summary of the model
print(model.summary())


                 Generalized Linear Model Regression Results                  
Dep. Variable:        No_of_Accidents   No. Observations:                    4
Model:                            GLM   Df Residuals:                        2
Model Family:                 Poisson   Df Model:                            1
Link Function:                    Log   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -23.134
Date:                Wed, 31 Jul 2024   Deviance:                       37.430
Time:                        16:26:46   Pearson chi2:                     27.0
No. Iterations:                     5   Pseudo R-squ. (CS):             0.3680
Covariance Type:            nonrobust                                         
                             coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------------------
Intercept                  2

## interpretation

The GLM analysis is focused on Main Line tracks’ ratio compared to Branch Line tracks regarding the number of accidents. The nominal variable line type is examined for its effect on the count of accidents using a Poisson distribution with a log link function.

Main Line
The coefficient for Line_Type[T. Main_Line] is 0. 5306. This means that, on the average, the Main Lines have the capability of terms of accident count compared to Branch Lines. More specifically, the annual rate of Main Lines Accidents is likely to hover at about forty-one. This is £3.80m or 1% lower than on Branch Lines. This is arrived at by taking the coefficient to the power of negative exponent as (e^-0. 5376=0. 589) indicating that Main Lines has a mean number of accidents that are about 59 percent of that on Branch Lines.

Branch Line
Branch Lines are the lowest level of categories in this model. Also the intercept value is equal to 2. 1401 is the logarithm of the anticipated accidents on Branch Lines. When thus exponentiated these translate to an expected number of approximately 8. 52 accidents on branch lines. This baseline estimate enables the participant Lines to be compared with Main Lines.


Summary
Main Lines are more favourable concerning the number of accidents that occur per year than Branch Lines with a decrease of about 41%. 1%. Nevertheless, the above reduction is not significant enough, and, therefore, it is impossible to claim that the accident rates of Main Line and Branch Line are significantly different based on the given dataset.

# Season

In [22]:
import pandas as pd
import numpy as np  # Make sure to import numpy
from sklearn.linear_model import PoissonRegressor
from sklearn.preprocessing import StandardScaler

# Load and prepare the data
file_path = 'peshawar.xlsx'
sheet_name = 'Season'
df = pd.read_excel(file_path, sheet_name=sheet_name)

# Replace spaces in column names with underscores
df.columns = df.columns.str.replace(' ', '_')

# Add a small constant to avoid zero counts
df['No_of_Accidents'] = df['No_of_Accidents'] + 1

# Convert 'Season' to a categorical variable
df['Season'] = df['Season'].astype('category')

# Prepare data for scikit-learn
X = pd.get_dummies(df['Season'], drop_first=True)
y = df['No_of_Accidents']

# Standardize features
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Fit Poisson regression model
model = PoissonRegressor(alpha=1.0)
model.fit(X_scaled, y)

# Extract coefficients
coefficients = model.coef_
intercept = model.intercept_

# Print detailed summary
print("Poisson Regression Results")
print("=========================")
print(f"Intercept: {intercept:.4f}")

# Map the coefficients to feature names
feature_names = pd.get_dummies(df['Season'], drop_first=True).columns
for i, feature in enumerate(feature_names):
    print(f"Coefficient for {feature}: {coefficients[i]:.4f}")

# Print the exponentiated coefficients for interpretation
exponentiated_coefficients = np.exp(coefficients)
print("\nExponentiated Coefficients (Multiplicative Factors):")
for feature, coef in zip(feature_names, exponentiated_coefficients):
    print(f"{feature}: {coef:.4f}")

# Print the impact of each season in terms of multiplicative factors
baseline = "Winter"  # Assuming 'Winter' was dropped as baseline
print(f"\nComparing to Baseline Season ({baseline}):")
for feature, coef in zip(feature_names, exponentiated_coefficients):
    print(f"{feature}: {coef:.4f} times the number of accidents compared to {baseline}")


Poisson Regression Results
Intercept: 1.9484
Coefficient for Spring: 0.0481
Coefficient for Summer: 0.2854
Coefficient for Winter: 0.4750

Exponentiated Coefficients (Multiplicative Factors):
Spring: 1.0492
Summer: 1.3303
Winter: 1.6079

Comparing to Baseline Season (Winter):
Spring: 1.0492 times the number of accidents compared to Winter
Summer: 1.3303 times the number of accidents compared to Winter
Winter: 1.6079 times the number of accidents compared to Winter


## interpretation 

We opted for Poisson regression instead of GLM due to practical issues such as perfect separation errors that can occur with GLM. Poisson regression in scikit-learn is straightforward for count data and avoids the complexities and potential fitting issues associated with GLMs in some cases. This approach simplifies the analysis while providing clear insights into the impact of different seasons on the number of accidents

The Intercept value is therefore 1. 9484 is the value of the quantity ‘accident rate for the baseline season on the log scale. This when exponentiated yields the base level of accidents for Winter season.

For the season-specific coefficients:

Next, the F statistic value for Spring is 0. 0481. Thus, log (number of accidents) indicating Spring is equal to 0. 1385 units higher than in Winter and within 0481 units of exceeding Summer. Raising this coefficient to the power of ex means predicting that the average number of accidents in Spring is approx 1. Thus, it has been found that the number of passengers during Spring is 0492 times that in Winter which shows a small rise.

Coefficient for Summer is zero. 2854. This implies that log number of Summer’s accidents is 0. 2854 units higher than in Winter season meaning that the higher the season the higher the probability of a network congestion. The resulting can be exponentiated to denote the fact that the estimated average of the number of accidents in Summer is, approximately 1. This constituted 3303 times that in Winter, a moderate rise.

The given figure of the season Winter being the starting point is equal to 1. It expanded 6079 times its size or to be better precise 6079 bigger than its initial size. This makes a confirmation that Winter is the reference category.

Therefore, the probability of an accident is somewhat more in Spring than the Winter- 1. 0492 times- more in Summer-1. 3303 times and the Winter is the standard or the rate to compare f accidents.

# Train Type

In [26]:
import pandas as pd
import statsmodels.api as sm

# Load the data
file_path = 'peshawar.xlsx'
sheet_name = 'Types of Train'
df = pd.read_excel(file_path, sheet_name=sheet_name)

# Replace spaces in column names with underscores
df.columns = df.columns.str.replace(' ', '_')

# Prepare data for GLM
# Define the dependent variable and independent variables
X = df[['Passenager', 'Good']]
y = df['No_of_Accidents']

# Add a constant to the independent variables matrix (for intercept)
X = sm.add_constant(X)

# Fit the GLM model with Poisson family
model = sm.GLM(y, X, family=sm.families.Poisson())
results = model.fit()

# Print the summary
print(results.summary())


                 Generalized Linear Model Regression Results                  
Dep. Variable:        No_of_Accidents   No. Observations:                    6
Model:                            GLM   Df Residuals:                        3
Model Family:                 Poisson   Df Model:                            2
Link Function:                    Log   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -9.9138
Date:                Wed, 31 Jul 2024   Deviance:                     0.075380
Time:                        16:44:44   Pearson chi2:                   0.0747
No. Iterations:                     4   Pseudo R-squ. (CS):             0.5458
Covariance Type:            nonrobust                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.2684      0.979      0.274      0.7

## interpretation

The Intercept value of the model is equal to 0. As 2 when there are no passenger or goods trains; and 2684 is the log of the expected number of accidents. This will help to establish the point of reference through which the impact of the other factors can be determined.

It was observed that for Passenager, the coefficient is 0. 2226. This implies that with each additional of the passenger train, there is an incremental change in the value of the log of the number of accidents of 0. 2226. This, when translated to a multiplicative factor means that each extra passenger train is accompanied by an increment of, roughly, twenty four. 9% increase in the expected number of accidents.PNG

Concerning the coefficient for Good, the result is equal to 0. 3094. This can be interpreted that for each additional goods train, the number of accidents as represented by the log number is raised by 0. 3094. Raising this figure to the power of goods train indicates that each additional goods train entails expenditure equivalent to about 36. Meaning an increase in the expected frequency of accidents by 2%. However, the coefficient of goods trains is insignificant implication that the effect of the goods trains on the number of accidents is less conclusive.

In general increasing the number of passenger trains shows that there is a strong relationship between the number of passenger trains and the number of accidents where for each increase in the number of these passenger trains the number of accidents increases notably. The impact of goods trains, however, cannot be easily determined mainly because the results are insignificant. By examining the model sum of squares value and comparing it to the residual sum of squares, one can say that the model bears a fair level of fit with data, which implies that the former accounts for a significant part of the variation in the numbers of the accidents.

# Cost of Damage

In [31]:
import pandas as pd
import statsmodels.api as sm

# Load the data
file_path = 'peshawar.xlsx'
sheet_name = 'Cost of Damage'
df = pd.read_excel(file_path, sheet_name=sheet_name)

# Replace spaces in column names with underscores
df.columns = df.columns.str.replace(' ', '_')

# Verify the updated column names
print(df.columns)  # Should now show the names with underscores

# Prepare data for GLM
# Define the dependent variable (Cost_of_Damage) and independent variables
X = df[['No_of_Accidents', 'No_Damge_to_PR', 'Cause_Damge_to_PR']]
y = df['Cost_of_Damage']

# Add a constant to the independent variables matrix (for intercept)
X = sm.add_constant(X)

# Fit the GLM model with Poisson family
model = sm.GLM(y, X, family=sm.families.Poisson())
results = model.fit()

# Print the summary
print(results.summary())


Index(['Year', 'No_of_Accidents', 'No_Damge_to_PR', 'Cause_Damge_to_PR',
       'Cost_of_Damage'],
      dtype='object')
                 Generalized Linear Model Regression Results                  
Dep. Variable:         Cost_of_Damage   No. Observations:                    6
Model:                            GLM   Df Residuals:                        3
Model Family:                 Poisson   Df Model:                            2
Link Function:                    Log   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:            -1.6681e+05
Date:                Wed, 31 Jul 2024   Deviance:                   3.3354e+05
Time:                        16:51:04   Pearson chi2:                 2.79e+05
No. Iterations:                     5   Pseudo R-squ. (CS):              1.000
Covariance Type:            nonrobust                                         
                        coef    std err          z      P>|z|      [0.025      0.975]
---

## interpretation

The Poisson regression model was applied to understand the relationship between the cost of damage and the independent variables: average of each accident occurrence, on the reputational loss, and the reason for reputational loss.

Coefficients Interpretation
Intercept: The intercept close to 13. The constant 9429 equals to log of the expected cost of damage with all the predictor variables equals to zero. This value defines the extent of the basic cost of damage.

No_of_Accidents: For No_of_Accidents we have -0 as the coefficient. 3727. This negative value tells that with the increase in the number of accidents the log of cost of damage has a tendency to decrease. In relative terms, which means that other things being equal the risk of a higher number of accidents is associated with a lower cost of the damages.

No_Damge_to_PR: The coefficient for No_Damge_to_PR is 0 As mentioned earlier. 2912. This is a positive value which will imply that as the level of damages to PR rises then the log of the cost of damage also rises. Therefore, any given level of damage is more costly to a PR, as relative to a NR, all other things equal.

Cause_Damge_to_PR: For Cause_Damge_to_PR, the coefficient is -0. 6639. Such negative coefficient gives information that the damage to PR is related to some extent decrease in the log of the cost of damage. Thus, some causes imply a lower cost of damage with other factors being equal.

Summary

Budget factor influence on cost of damage is clearly illustrated in the developed model. It is also Rather, the number of accidents appears to be negatively related to the cost while on the other hand, damage to PR seems to have a positive relationship with the variable. The cause of the damage of L2 in relation to PR becomes cost negative. Looking at the presented equations, one can see that the model has high Pseudo R-squared and the coefficients are allmited sample.

# Severity

In [2]:
import pandas as pd
import statsmodels.api as sm

# Load the data
file_path = 'peshawar.xlsx'
sheet_name = 'Severity'
df = pd.read_excel(file_path, sheet_name=sheet_name)

# Rename columns to remove extra spaces and make them consistent
df.columns = df.columns.str.strip().str.replace(' ', '_')

# Verify column names after renaming
# Define dependent variables and independent variable
X = df[['No_of_Accidents']]
y_casualties = df['No_Casulties']
y_deaths = df['No_of_Deaths']
y_injuries = df['No_of_Injuries']

# Add a constant to the independent variables matrix (for intercept)
X = sm.add_constant(X)

# Fit GLM model for each severity measure
model_casualties = sm.GLM(y_casualties, X, family=sm.families.Poisson()).fit()
model_deaths = sm.GLM(y_deaths, X, family=sm.families.Poisson()).fit()
model_injuries = sm.GLM(y_injuries, X, family=sm.families.Poisson()).fit()

# Print the summary for each model
print("Model Summary for Number of Casualties:")
print(model_casualties.summary())

print("\nModel Summary for Number of Deaths:")
print(model_deaths.summary())

print("\nModel Summary for Number of Injuries:")
print(model_injuries.summary())


Model Summary for Number of Casualties:
                 Generalized Linear Model Regression Results                  
Dep. Variable:           No_Casulties   No. Observations:                    6
Model:                            GLM   Df Residuals:                        4
Model Family:                 Poisson   Df Model:                            1
Link Function:                    Log   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -9.0807
Date:                Wed, 31 Jul 2024   Deviance:                      0.20388
Time:                        19:51:02   Pearson chi2:                    0.200
No. Iterations:                     4   Pseudo R-squ. (CS):             0.2723
Covariance Type:            nonrobust                                         
                      coef    std err          z      P>|z|      [0.025      0.975]
-----------------------------------------------------------------------------------
co

## interpretation 
Number of Casualties Model

The GLM analysis for predicting the number of casualties reveals an intercept of 0.3582. This suggests that when there are no accidents, the expected number of casualties is roughly zero, although this is not statistically significant. The coefficient for the number of accidents is 0.1655, indicating that each additional accident is associated with approximately a 16.55% increase in the number of casualties. However, this relationship is not statistically significant, as indicated by the p-value of 0.169. The Pseudo R-squared value of 0.2723 suggests a moderate level of fit, meaning that the model explains a portion of the variability in the number of casualties based on the number of accidents.

Number of Deaths Model

For the number of deaths, the intercept is -1.0771, which is not statistically significant, with a p-value of 0.316. The coefficient for the number of accidents is 0.2944, suggesting a 29.44% increase in the number of deaths for each additional accident. This coefficient has a p-value of 0.109, which implies that while the effect is positive, it is not statistically significant at conventional levels. The Pseudo R-squared value of 0.3686 indicates that this model has a somewhat better fit compared to the casualties model, explaining more of the variability in the number of deaths based on the number of accidents.

Number of Injuries Model

In the model for predicting the number of injuries, the intercept is 1.2700, which is statistically significant with a p-value of 0.009, suggesting a baseline level of injuries. The coefficient for the number of accidents is 0.0733, indicating a smaller increase of about 7.33% in the number of injuries for each additional accident. This effect is not statistically significant, with a p-value of 0.442. The Pseudo R-squared value of 0.09346 shows that this model explains only a small proportion of the variance in the number of injuries.

Overall Interpretation

The combined analysis shows that while there is a positive association between the number of accidents and each type of severity (casualties, deaths, and injuries), the strength and significance of these relationships vary. The models indicate that the number of accidents does have some impact on casualties, deaths, and injuries, but the effects are not consistently strong or statistically significant. This suggests that additional factors or more complex interactions might be influencing these outcomes, and further exploration is needed to improve the models and understand the full scope of factors affecting railway accident severity.